### PR TITLE
Add `--target` API option

### PR DESCRIPTION
**Pull request by lazar-mitrovic**
This PR adds `--target` API option which is used to specify compilation target platform for `native-image`

Usage: 
```
native-image --target=linux-amd64 HelloWorld
```

### CODE REVIEW

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/SubstrateOptions.java [hunk 1]
```diff
@@ -75,6 +75,10 @@
     @Option(help = "Build statically linked executable (requires static libc and zlib)")//
     public static final HostedOptionKey<Boolean> StaticExecutable = new HostedOptionKey<>(false);
 
+    @APIOption(name = "target")//
+    @Option(help = "Selects native-image compilation target (in `os-arch` format)")//
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/SubstrateOptions.java:79]: in the `<OS>-<architecture>` format. Also, define what is the default.

File: substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java [hunk 2]
```diff
@@ -1212,6 +1215,36 @@ private static String getAgentOptions(String options, String optionName) {
         return Arrays.stream(options.split(",")).map(option -> optionName + "=" + option).collect(Collectors.joining(","));
     }
 
+    private List<String> getTargetArguments() {
+        /*
+         * Since regular hosted options are parsed at later phase of NativeImageGeneratorRunner
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java:1220]: at a later phase

File: substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java [hunk 3]
```diff
@@ -1212,6 +1215,36 @@ private static String getAgentOptions(String options, String optionName) {
...
+         * process (see comments for NativeImageGeneratorRunner.defaultPlatform), we are parsing
+         * --target argument here, and generating required internal arguments.
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java:1222]: the --target argument

**Comment by vjovanov** [comment]
For this PR to be complete, it would be necessary to catch unsupported targets and report errors in an actionable way.

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/SubstrateOptions.java (hunk 1 above)
Comment (lazar-mitrovic) [reply substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/SubstrateOptions.java]: fixed

File: substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java (hunk 3 above)
Comment (lazar-mitrovic) [reply substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java]: Fixed

File: substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java (hunk 2 above)
Comment (lazar-mitrovic) [reply substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java]: Fixed

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/SubstrateOptions.java [hunk 4]
```diff
@@ -75,6 +75,10 @@
     @Option(help = "Build statically linked executable (requires static libc and zlib)")//
     public static final HostedOptionKey<Boolean> StaticExecutable = new HostedOptionKey<>(false);
 
+    @APIOption(name = "target")//
+    @Option(help = "Selects native-image compilation target (in <OS>-<architecture> format)")//
+    public static final HostedOptionKey<String> TargetPlatform = new HostedOptionKey<>(System.getProperty("os.name") + "-" + System.getProperty("os.arch"));
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/SubstrateOptions.java:80]: I thought that you say in the message what is the default. And the value should stay as "".

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/SubstrateOptions.java [hunk 5]
```diff
@@ -75,6 +75,10 @@
     @Option(help = "Build statically linked executable (requires static libc and zlib)")//
     public static final HostedOptionKey<Boolean> StaticExecutable = new HostedOptionKey<>(false);
 
+    @APIOption(name = "target")//
+    @Option(help = "Selects native-image compilation target (in <OS>-<architecture> format). If isn't set, defaults to host's OS-architecture pair.")//
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/SubstrateOptions.java:79]: `If isn't set` is superfluous

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/NativeImageGenerator.java [hunk 6]
```diff
@@ -367,6 +392,33 @@ public static Platform defaultPlatform(ClassLoader classLoader) {
...
+         * appropriate interfaces for OS and Architecture which would then be implemented by all
+         * Platform classes.
+         */
+        Class<?>[] interfaces = p.getClass().getInterfaces();
+        for (Class<?> i : interfaces) {
+            if (InternalPlatform.PLATFORM_JNI.class.isAssignableFrom(i)) {
+                return i.getSimpleName();
+            }
+        }
+        VMError.shouldNotReachHere("Couldn't find OS for specified platform. Check if provided Platform class implements PLATFORM_JNI.");
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/NativeImageGenerator.java:411]: throw should go here and return is then not necessary.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/NativeImageGeneratorRunner.java [hunk 7]
```diff
@@ -169,7 +170,17 @@ public static ImageClassLoader installNativeImageClassLoader(String[] classpath,
...
-        return new ImageClassLoader(NativeImageGenerator.defaultPlatform(nativeImageClassLoader), nativeImageClassLoaderSupport);
+        Platform targetPlatform = NativeImageGenerator.getTargetPlatform(nativeImageClassLoader);
+        Platform hostPlatform = NativeImageGenerator.getHostPlatform();
+
+        if (System.getProperty("svm.targetPlatform") != null) {
+            /*
+             * This check is only meaningful for default platforms that are supported by --target
+             * parameter (see comment in NativeImageGenerator.getPlatformOS for more information)
+             */
+            UserError.guarantee(NativeImageGenerator.isSameOS(targetPlatform, hostPlatform), "Cross OS compilation isn't supported at the moment.");
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/NativeImageGeneratorRunner.java:181]: What about android? Shouldn't we just check if libs are present and say target not supported?

Comment (lazar-mitrovic) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/NativeImageGeneratorRunner.java:181]: As for now, we are building Android as `LINUX_AARCH64`. My reasoning is that when Android gets implemented as a `Platform` (2 PR's down the line), it should be extended from `LINUX`, which would still satisfy this check.

Checking for static libraries is already done in `NativeLibraries.java:336`

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/NativeImageGenerator.java (hunk 6 above)
Comment (lazar-mitrovic) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/NativeImageGenerator.java]: Fixed.

File: substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/SubstrateOptions.java (hunk 5 above)
Comment (lazar-mitrovic) [reply substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/SubstrateOptions.java]: Fixed.

File: substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java [hunk 8]
```diff
@@ -1212,6 +1224,42 @@ private static String getAgentOptions(String options, String optionName) {
...
+        String arch = parts[1];
+
+        if (isListArgumentSet(customJavaArgs, "-D" + Platform.PLATFORM_PROPERTY_NAME)) {
+            NativeImage.showWarning("Usage of -D" + Platform.PLATFORM_PROPERTY_NAME + " might conflict with --target parameter.");
+        }
+
+        if (!isListArgumentSet(customJavaArgs, "-Dsvm.targetPlatform")) {
+            customJavaArgs.add("-Dsvm.targetPlatform=" + target);
+        }
+        if (!isListArgumentSet(customJavaArgs, "-Dsvm.targetName")) {
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java:1255]: Should we remove those two?

File: sdk/src/org.graalvm.nativeimage/src/org/graalvm/nativeimage/Platform.java [hunk 9]
```diff
@@ -102,7 +131,36 @@ static boolean includedIn(Class<? extends Platform> platformGroup) {
...
 
+    /*
+     * Operating system bases.
+     */
+    /**
+     * Interface encapsulating all linux-derived operating systems.
+     *
+     * @since 20.3
+     */
+    interface LINUX_BASE extends InternalPlatform.PLATFORM_JNI {
```
Comment (vjovanov) [inline sdk/src/org.graalvm.nativeimage/src/org/graalvm/nativeimage/Platform.java:147]: Should these be public?

File: sdk/src/org.graalvm.nativeimage/snapshot.sigtest [hunk 10]
```diff
@@ -228,56 +228,56 @@ CLSS public abstract interface static org.graalvm.nativeimage.Platform$AMD64
  outer org.graalvm.nativeimage.Platform
 intf org.graalvm.nativeimage.Platform
 
-CLSS public abstract interface static org.graalvm.nativeimage.Platform$DARWIN
+CLSS public abstract interface static org.graalvm.nativeimage.Platform$DARWIN_BASE
```
Comment (vjovanov) [inline sdk/src/org.graalvm.nativeimage/snapshot.sigtest:231]: I think this should stay the same.

File: substratevm/src/com.oracle.svm.core.posix/src/com/oracle/svm/core/posix/darwin/DarwinSystemPropertiesSupport.java [hunk 11]
```diff
@@ -60,6 +61,12 @@ protected String tmpdirValue() {
         }
     }
 
+    @Override
+    protected String osNameValue() {
+        return Platform.includedIn(Platform.IOS.class)
+                        ? "iOS" : "Mac OS X";
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.core.posix/src/com/oracle/svm/core/posix/darwin/DarwinSystemPropertiesSupport.java:67]: This can be put on one line.

**Comment by vjovanov** [comment]
Please reformat the code.

**Pull request closed without merging**